### ISSUE TITLE
completion ignoring . prefixed arguments

### ISSUE THREAD

**Issue by nhooyr** [fish-shell/fish-shell#3707]
- [x] Have you checked if problem occurs with [fish 2.4.0](/fish-shell/fish-shell/releases/tag/2.4.0)?
- [ ] Tried fish without third-party customizations *(check `sh -c 'env HOME=$(mktemp -d) fish'`)*?

**fish version installed** *(`fish --version`)*: 2.4.0

**OS/terminal used**:  macOS 10.12.2 iTerm 3.0.13

completion ignores arguments that start with `.`. I think this is to ignore dotfiles but I have a command that can be completed with a file that starts with `.` and it should not be ignored. Is it possible to disable this?

watch https://asciinema.org/a/9wnu3jy22hxfjf4au18m7ob29

**Comment by ridiculousfish** [comment]
This is by design. The default file completion will only show hidden files if your argument starts with a `.` After all, hidden files should be hidden.

We wouldn't introduce a config option for this.

I thought it should be possible to write a custom completion, but it looks like fish squashes custom completions that start with `.` unless the token starts with `.` (I thought this only happened for file completions, but it looks like it happens all the time). That's probably a bug, and one that can be fixed.

For now though I think this just isn't possible, sorry!

**Comment by nhooyr** [comment]
Yea, I'm asking about custom completion not for file completion.

**Comment by krader1961** [comment]
@nhooyr: Please update this issue with your custom completion script that is not working as you desire. I cannot tell from your asciinema recording what `complete` commands were in effect that you expected to behave differently. Also, your asciinema recording is ambiguous since it has insufficient context, in conjunction with this issue, to understand what is not working as you expected.

It should be possible to inhibit the usual file name completion for strings beginning with `.`. But that is clearly an enhancement and we need more details about your use case.

**Comment by nhooyr** [comment]
Sorry, I didn't realize I was so unclear when I wrote this.

Here is the script https://github.com/nhooyr/dotfiles/blob/812056d8e13d99dd6bb3eced79b8d13dbb8e91d5/.config/fish/completions/bd.fish

Fish doesn't show me `.config .foo .bar` if I don't have a period in front when I attempt to complete. It only shows me `foo 99j bar`.

**Comment by krader1961** [comment]
Hmmm, okay, you're adding an explicit completion that includes arguments that begin with a period:

```
complete -c bd -f -a '.config foo 99j .foo .bar bar'
```

And those do not show up in the completion list. I'm inclined to agree that explicitly enumerating arguments that begin with a period should be displayed and selectable after pressing [tab] even if you haven't typed a period. Whereas the behavior for file name expansion should continue to exclude such file names unless you have started the argument with a period before pressing [tab].

**Comment by ridiculousfish** [comment]
Yes, me too - the hidden file behavior is only supposed to apply to files, not bespoke completions.

**Comment by rmartine-ias** [comment]
# I would like this too
This would be a useful enhancement to me -- my company has a practice of having CI code in a `.ci/` directory, and I frequently make changes to it. `git add` does not tab complete the files in that directory. It [looks like](https://github.com/fish-shell/fish-shell/blob/master/share/completions/git.fish#L1100) it should:

```fish
complete -f -c git -n '__fish_git_using_command add' -a '(__fish_git_files modified untracked deleted unmerged modified-staged-deleted)'
```

(`__fish_git_files` correctly returns the dot-prefixed paths. I think it would make sense, like you said, for explicitly added completions to work, even if they start with a dot.)

Each time I do a tab completion for this, I get confused for a second, go "oh right", and add a dot and press tab again. It's a minor annoyance, but one I deal with very frequently.

<hr>

## Temp. solution for git add specifically

Now that I've figured out it's fish's fault, and not a misconfigured git, I can do something about it. For anyone else with this same exact `git add` issue, here's *a* solution (excuse my poor fish):
```fish

function confirm -a 'prompt' -d "Confirm whether or not to do something"
    if test -z "$prompt"
        set prompt 'Are you sure?'
    end

    read -f -p "set_color green; echo \"$prompt [y/N]\"; set_color normal" response
    switch "$response"
        case y Y
            true
        case '*'
            false
    end
end

function gap -d "Git add one file, or prompt for all"
    if [ (git status --porcelain | wc -l) = 1 ]
        git add -v -A # -v prints the added file
    else
        # can we prompt one by one, but better than git add --interactive?
       git status; confirm "Add all?"; and git add -A
    end
end
```

Replacing `ga <tab><return>` with `gap<return>` is.. better. Remember as "git add, please".

**Comment by ridiculousfish** [comment]
Fixed in b7de768c73f0a9b0a097b83db1f60726c3a9661a. Thank you for filing.

**Comment by bricker** [comment]
I know this is an old thread, but I want to express my opinion on this:

> After all, hidden files should be hidden.

I understand that historically dotfiles have been considered "hidden", but the reality is that the dotfile pattern is way over-used and has lost any concrete meaning.

For example, here are some common files/directories that use the leading-dot convention, but are manually edited or accessed often and have no reason to be hidden in a list of files:

- ~/.bashrc
- ~/.zshrc
- ~/.vimrc
- ~/.config/*
- ~/.ssh/*
- .eslintrc
- .gitignore
- .github/*

The `fish-shell` repository itself has several dotfiles that have no reason to be hidden.

- .cirrus.yml
- .clang-format
- .clang-tidy
- .oclint
- .editorconfig

But, those are all mixed in with other files/directories that _aren't_ manually edited or accessed, such as:

- ~/.bash_history
- ~/.cache
- ~/.python_history
- .mypy_cache
- .build
- .git
- .node_modules

So, a file/directory with a leading dot might mean:

- not a source code file (generally in a codebase)
- a configuration file in `$HOME`
- temporary/transient data in `$HOME` (eg cache, logs, command history)
- anything else that the owner wants it to mean

In conclusion: I believe that hiding dotfiles is incorrect, because it's not a defined standard. So I think there should at least be a configuration to ignore the convention and allow tab-complete to show _all_ files in a directory. Even the standard `ls` command has an option to list _all_ files.

**Comment by miyl** [comment]
Very good points, @bricker.

Glad to see the bug was fixed. I hope adding a configuration option may be considered now, 6 years later.

**Comment by retrobit** [comment]
Not regarding completions, but simple directory traversal, @ridiculousfish mentions here: 

https://github.com/fish-shell/fish-shell/commit/b28eae8be9dfff290cddadf8c54f43cacad49cc5

that

> "By default, fish does not complete files that have leading dots, unless the
wildcard itself has a leading dot."

Does this mean there is configuration to change this default behavior? If so, where can I find it? I have been searching which led me to this related (but not identical) issue thread.
